ext/standard/tests/file/file_descriptor_non_castable_user_stream.phpt

```diff
@@ -0,0 +1,79 @@
+--TEST--
+Test file_descriptor() function: error on uncastable user stream
+--FILE--
+<?php
+
+// TODO Improve by mimicking tests from GH 10173 PR
+class DummyStreamWrapper
+{
+    /** @var resource|null */
+    public $context;
+
+    /** @var resource|null */
+    public $handle;
+
+    public function stream_cast(int $castAs)
+    {
+        return false;
+    }
+
+    public function stream_close(): void { }
+
+    public function stream_open(string $path, string $mode, int $options = 0, ?string &$openedPath = null): bool
+    {
+        return true;
+    }
+
+    public function stream_read(int $count)
+    {
+        return 0;
+    }
+
+    public function stream_seek(int $offset, int $whence = SEEK_SET): bool
+    {
+        return true;
+    }
+
+    public function stream_set_option(int $option, int $arg1, ?int $arg2): bool
+    {
+        return false;
+    }
+
+    public function stream_stat()
+    {
+        return [];
+    }
+
+    public function stream_tell()
+    {
+        return [];
+    }
+
+    public function stream_truncate(int $newSize): bool
+    {
+        return true;
+    }
+
+    public function stream_write(string $data) { }
+
+
+    public function unlink(string $path): bool
+    {
+        return false;
+    }
+}
+stream_wrapper_register('custom', DummyStreamWrapper::class);
+
+$fp = fopen("custom://myvar", "r+");
+try {
+    var_dump(file_descriptor($fp));
+} catch (\TypeError $e) {
+    echo $e->getMessage(), PHP_EOL;
+}
```

Comment:
not sure if SPL exceptions can be used here, but `InvalidArgumentException` makes more sense than `TypeError` as `$fp` is of the correct type ( `resource` ), but it's not a valid resource to be used with `file_descriptor`.

```suggestion
try {
    var_dump(file_descriptor($fp));
} catch (\InvalidArgumentException $e) {
    echo $e->getMessage(), PHP_EOL;
}
```

Reply:
We use TypeError for invalid types of resources else where. Moreover, the reason for that is ultimately the objective is to get rid of resources and convert them to objects, where different resource types will be different objects.

SPL exceptions are also rubbish and should not be used within ext/standard.

Reply:
yea, I think a better solution for the future would be to introduce an interface for resource-like objects with `getFileDescriptor(): ?int` method, and this function would simply call that. returning `null` if the resource is in an invalid state or something like that.

Reply:
Moving streams to objects is going to be a massive project where all those details can (and should) be fleshed out :D